Sources/NIO/CircularBuffer.swift

```diff
@@ -28,61 +28,80 @@ public struct CircularBuffer<Element>: CustomStringConvertible, AppendableCollec
     private var buffer: ContiguousArray<Element?>
 
     /// The index into the buffer of the first item
-    private(set) /* private but tests */ internal var headIdx = 0
+    private(set) /* private but tests */ internal var headIdx: Index
```

Comment:
I think it would be enough to just store the backing index as `Int` here. We can always fabricate the `startIndex` as actual `Index` when we need to as `Index(backingIndex self.headIdx, count: self.backing.count, isGEQ: true)` right?

Reply:
Well mostly because those are 2 main indexes of the `CircularBuffer` so it made sense they are of `Index` type since all operations are now performed over that type.
I would feel like seeing something that someone didn't intend working that way in this code if those were `Int` and constructed on access:
` /// Return all valid indices of the ring.
    public var indices: RangeType<Index> {
        return self.startIndex ..< self.endIndex
    }
    /// Returns the index of the first element of the ring.
    public var startIndex: Index {
        return self.headIdx
    }

    /// Returns the ring's "past the end" position -- that is, the position one greater than the last valid subscript argument.
    public var endIndex: Index {
        return self.tailIdx
    }`
Additionally, masking would need to be introduced into `CircularBuffer` for `count`, and it's contained inside `Index` at the moment.
The rest of the code relies on `count`, `startIndex` and `endIndex` and maybe less important for this question but many operations move head and tail by 1 or -1. All of this seems very unpredictable performance wise compared to storing them, but I'm

Reply:
@JovanMilenkovic the downside of storing those as `Index` is that it makes `CircularBuffer` a lot wider. If we stored them as `Int`, that's two words ( 1 for startIndex, 1 for endIndex). Storing them as `Index` would be 6 words ((2 words + 1 byte) for startIndex + padding + (2 words + 1 byte) for endIndex + padding). So I'd advise to only store what we need to store and calculate the rest